p2p/pex/pex_reactor.go

```diff
@@ -464,45 +464,116 @@ func (r *Reactor) ensurePeers() {
-	maxAttempts := numToDial * 3
+	maxToDialAttempts := numToDial * 3
 
-	for i := 0; i < maxAttempts && len(toDial) < numToDial; i++ {
-		if !r.IsRunning() || !r.book.IsRunning() {
-			return
-		}
+	// Calculate reserve size as 50% of numToDial with a minimum of 10
+	reserveSize := cmtmath.MaxInt(numToDial/2, 10)
+	maxReserveAttempts := reserveSize * 3
 
-		try := r.book.PickAddress(newBias)
-		if try == nil {
-			continue
+	filter := func(ka *knownAddress) bool {
+		attempts, lastDialedTime := r.dialAttemptsInfo(ka.Addr)
+		if r.TooEarlyToDial(ka.Addr, attempts, lastDialedTime) != nil {
+			return false
+		}
+		if r.MaxAttemptsToDial(ka.Addr, attempts) != nil {
+			r.book.MarkBad(ka.Addr, defaultBanTime)
+			return false
+		}
+		if r.Switch.IsDialingOrExistingAddress(ka.Addr) {
+			return false
+		}
+		if _, selected := toDial[ka.Addr.ID]; selected {
+			return false
+		}
+		if _, selected := reserve[ka.Addr.ID]; selected {
+			return false
 		}
-		if _, selected := toDial[try.ID]; selected {
+		return true
+	}
+
+	for i := 0; i < maxToDialAttempts && len(toDial) < numToDial; i++ {
+		prospectivePeer := r.book.PickAddress(newBias, filter)
+		if prospectivePeer == nil {
 			continue
 		}
-		if r.Switch.IsDialingOrExistingAddress(try) {
+		toDial[prospectivePeer.ID] = prospectivePeer
+	}
+
+	for i := 0; i < maxReserveAttempts && len(reserve) < reserveSize; i++ {
+		prospectivePeer := r.book.PickAddress(newBias, filter)
+		if prospectivePeer == nil {
 			continue
 		}
-		// TODO: consider moving some checks from toDial into here
-		// so we don't even consider dialing peers that we want to wait
-		// before dialing again, or have dialed too many times already
-		toDial[try.ID] = try
+		reserve[prospectivePeer.ID] = prospectivePeer
 	}
 
+	toDialCount := len(toDial)
+	reserveCount := len(reserve)
+
+	var (
+		successCh = make(chan int)
+		errorCh   = make(chan int)
+		wg        sync.WaitGroup
+	)
+
+	// Dial picked addresses
 	// Dial picked addresses
 	for _, addr := range toDial {
+		wg.Add(1)
 		go func(addr *p2p.NetAddress) {
+			defer wg.Done()
 			err := r.dialPeer(addr)
 			if err != nil {
-				switch err.(type) {
-				case ErrMaxAttemptsToDial, ErrTooEarlyToDial:
-					r.Logger.Debug(err.Error(), "addr", addr)
-				default:
-					r.Logger.Debug(err.Error(), "addr", addr)
+				errorCh <- 1
+				r.Logger.Debug(err.Error(), "addr", addr)
+				// If reserve addresses are available, try to dial them due to this toDial address erroring out
+				for id, reserveAddr := range reserve {
+					if reserveAddr != nil {
+						delete(reserve, id)
+						err := r.dialPeer(reserveAddr)
+						if err != nil {
+							errorCh <- 1
+							r.Logger.Debug(err.Error(), "addr", addr)
+						} else {
+							successCh <- 1
+						}
+						if err == nil {
+							break
+						}
+					}
 				}
+			} else {
+				successCh <- 1
 			}
 		}(addr)
 	}
 
+	go func() {
+		wg.Wait()
+		close(successCh)
+		close(errorCh)
+	}()
+
+	go func() {
```

Comment:
I have the impression that, 30 seconds after, when we run `ensurePeers` again, this method could still be running. Not ideal...